cloudinit/distros/__init__.py

```diff
@@ -908,6 +910,50 @@ def lock_passwd(self, name):
             util.logexc(LOG, "Failed to disable password for user %s", name)
             raise e
 
+    def unlock_passwd(self, name):
+        """
+        Unlock the password of a user, i.e., enable password logins
+        """
+        # passwd must use short '-u' due to SLES11 lacking long form '--unlock'
+        unlock_tools = (["passwd", "-u", name], ["usermod", "--unlock", name])
+        try:
+            cmd = next(tool for tool in unlock_tools if subp.which(tool[0]))
+        except StopIteration as e:
+            raise RuntimeError(
+                "Unable to unlock user account '%s'. No tools available. "
+                "  Tried: %s." % (name, [c[0] for c in unlock_tools])
+            ) from e
+        try:
+            _, err = subp.subp(cmd, rcs=[0, 3])
+        except Exception as e:
+            util.logexc(LOG, "Failed to enable password for user %s", name)
+            raise e
+        if err:
+            # if "passwd" or "usermod" are unable to unlock an account with
+            # an empty password then they display a message on stdout. In
+            # that case then instead set a blank password.
+            passwd_set_tools = (
+                ["passwd", "-d", name],
+                ["usermod", "--password", "''", name],
+            )
```

Comment:
I'm very hesitant to call this behavior as fallback if `lock_passwd: False` as it is a poor security policy decision which is why the passwd tooling exits non-zero (3) to warn about this and be noisy.  I'm not thrilled about cloud-init trying to support this lax security policy, but I can think of a couple of cases where it may make sense to have a completely open set of instances on a private network to simplify tooling and automation.

I think minimally cloud-init should:
- log a warning about this behavior which will allow cloud-init status to exit non-zero(2) telling the admin that they have an "poor security practice open system user account with an empty password defined" if someone configures their user with lock_passwd false and doesn't provide a valid non-empty passwd/plain_text_passwd

Ideally, if we think a person really wants to force an empty password and completely expose that instance to simple remote console attacks without getting a warning,  then we should probably avoid calling this `passwd -d` unless either an empty  `plain_text_passwd: ""` or `passwd: ""`  was explicitly provided. Otherwise, we risk opening up password-less access for folks who didn't realize the impact of providing ` lock_passwd: false`  without some form of explicit passwd defined.

I'd be a little more convinced to support this password-less account if it requires explicit additional configuration (such as providing an empty string to `*passwd` cloud-config user keys in user-data to define "no I really do want empty passwords for this use, I know the risks"


So ideal way I'd like to allow folks to express this would be:
```
#cloud-config 
users:
- default
- name: openuser
  gecos: Open User with no password
  primary_group: openuser
  lock_passwd: false
  passwd: ""  
  # or
  plain_text_passwd: ""
```

Reply:
I have addressed this now.